src/source-lib.c

```diff
@@ -0,0 +1,180 @@
+ *
+ * \param tv                    Pointer to the per-thread structure.
+ * \param data                  Pointer to the context.
+ * \return                      Error code.
+ */
+TmEcode DecodeLibThreadDeinit(ThreadVars *tv, void *data)
+{
+    if (data != NULL)
+        DecodeThreadVarsFree(tv, data);
+
+    time_set = false;
+    SCReturnInt(TM_ECODE_OK);
+}
+
+/** \brief main decoding function.
+ *
+ *  This method receives a packet and tries to identify layer 2 to 4 layers.
+ *
+ * \param tv                    Pointer to the per-thread structure.
+ * \param p                     Pointer to the packet.
+ * \param data                  Pointer to the context.
+ * \return                      Error code.
+ */
+TmEcode DecodeLib(ThreadVars *tv, Packet *p, void *data)
+{
+    SCEnter();
+    DecodeThreadVars *dtv = (DecodeThreadVars *)data;
+
+    BUG_ON(PKT_IS_PSEUDOPKT(p));
+
+    /* update counters */
+    DecodeUpdatePacketCounters(tv, dtv, p);
+
+    /* If suri has set vlan during reading, we increase vlan counter */
+    if (p->vlan_idx) {
+        StatsIncr(tv, dtv->counter_vlan);
+    }
+
+    /* call the decoder */
+    DecodeLinkLayer(tv, dtv, p->datalink, p, GET_PKT_DATA(p), GET_PKT_LEN(p));
+
+    PacketDecodeFinalize(tv, dtv, p);
+
+    SCReturnInt(TM_ECODE_OK);
+}
+
+/** \brief process a single packet.
+ *
+ * \param tv                    Pointer to the per-thread structure.
+ * \param data                  Pointer to the raw packet.
+ * \param datalink              Datalink type.
+ * \param ts                    Timeval structure.
+ * \param len                   Packet length.
+ * \param tenant_id             Tenant id of the detection engine to use.
+ * \param flags                 Packet flags (packet checksum, rule profiling...).
+ * \param iface                 Sniffing interface this packet comes from (can be NULL).
+ * \return                      Error code.
+ */
+int TmModuleLibHandlePacket(ThreadVars *tv, const uint8_t *data, int datalink, struct timeval ts,
+        uint32_t len, uint32_t tenant_id, uint32_t flags, const char *iface)
+{
+
+    /* If the packet is NULL, consider it as a read timeout. */
+    if (data == NULL) {
+        TmThreadsSetFlag(tv, THV_CAPTURE_INJECT_PKT);
+        TmThreadsCaptureHandleTimeout(tv, NULL);
+        SCReturnInt(TM_ECODE_OK);
+    }
+
+    Packet *p = PacketGetFromQueueOrAlloc();
+    if (unlikely(p == NULL)) {
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+
+    /* If we are processing a PCAP and it is the first packet we need to set the timestamp. */
+    SCTime_t timestamp = SCTIME_FROM_TIMEVAL(&ts);
+    if (!time_set && !TimeModeIsLive()) {
+        TmThreadsInitThreadsTimestamp(timestamp);
+        time_set = true;
+    }
+
+    PKT_SET_SRC(p, PKT_SRC_WIRE);
+    p->ts = timestamp;
+    p->datalink = datalink;
+    p->tenant_id = tenant_id;
+    p->flags |= flags;
+
+    /* Set the sniffing interface. */
+    if (iface) {
+        p->livedev = LiveGetDevice(iface);
```

Comment:
this seems like something you'd want to avoid per packet, so should we pass in the livedev ptr directly?

Reply:
I think it would make sense to require the pointer to be passed in directly.

But really, the concept of "live device" and library mode doesn't seem to make sense.  The packets could come from anywhere.  What in the "LiveDevice" structure is really needed?  Seems it's required for multi-tenant.  Might be a naming thing, would it have been named `Interface` I probably wouldn't give it a second thought :)

Reply:
Yes the idea is that, unlike other capture modes, a worker thread is not bound to an interface. The same worker thread could process packets coming from different interfaces. But I agree this could be provided in the API call directly.

Reply:
I would suspect that some lib users would still want to pass in a "interface" abstraction, which for now is Livedev. I agree it would be better to rename it to something like SCInterface or something. But that is for later.